### PR TITLE
Revert "[Agent] Remove heartbeat from bundled beats in agent."

### PR DESCRIPTION
**Pull request by shahzad31**
Adding heartbeat into agent, since we are restarting work on fleet integration with uptime/heartbeat in 7.11

Reverts elastic/beats#21602

### CODE REVIEW

**Comment by blakerouse** [comment]
I think this brings back into the question if heartbeat should be bundled. With the change it will be bundled.

I think its best for it to be bundled, but that is up for discussion.

@ph @ruflin ^

**Comment by shahzad31** [comment]
@elasticmachine merge upstream

**Comment by michalpristas** [comment]
if this is just to ease up development on apm side and not a product wise decision i think it would be better to focus on this issue instead #19623

**Comment by ruflin** [comment]
We need to have the bundling discussion again. I'm suggesting to first add support for heartbeat and not bundle it to unblock @shahzad31 and we will have a follow up discussion on bundling it or not as a follow up.

**Comment by paulb-elastic** [comment]
There is a related [discussion issue](https://github.com/elastic/synthetics/issues/140) open regarding the packaging of Synthetics/Heartbeat/Agent

**Comment by andrewvc** [comment]
As a note on testing, in the `x-pack/elastic-agent` folder, you can run:

`env DEV=true PLATFORMS="+all linux/amd64" mage package` will build a linux version of this in: `x-pack/elastic-agent/build/distributions`.

Use `darwin/amd64` if you're on a mac.

You can unpack the `tar.gz` in `build/distributions` and add a monitor to the `elastic-agent.yml` file using the heartbeat syntax, making sure to prefix the `type` config option with `synthetics/` so, for `http` you'd use:

```
inputs:
  - type: synthetics/http
    id: blahblah
    name: Blah
    urls: "https://www.elastic.co"
    schedule: "@every 15s"
```

**Comment by ruflin** [comment]
@michalpristas @blakerouse Seems something around fleet e2e tests is failing here.
@urso Could you take a look?

Overall LGTM to get it bundled in.

**Review by blakerouse — APPROVED** [review]
Looks good.

**Comment by urso** [comment]
I presume this is derailing discussion on this PR, but Andrew gave this configuration example:

```
inputs:
  - type: synthetics/http
    id: blahblah
    name: Blah
    urls: "https://www.elastic.co"
    schedule: "@every 15s"
```

For consistency with the metricbeat and filebeat integration, should we require to have streams?

Metricbeat based inputs use the pattern `*/metrics` but for heartbeat we use `synthetics/*`? Which one is the naming pattern we should standardize on?

**Comment by ruflin** [comment]
I would leave out the `streams` part as it seems no grouping / sharing of config options is needed for the synthetics use case.

Good point about the metrics / synthetics pre / post fix. The reason I prefer system/metrics system/logs is that it felt more intuitive to me to write and easier to remember. But this is very subjective I guess. On synthetics, I think both could work but no strong preference. I'm more worried now about not being consistent.

**Comment by urso** [comment]
> I would leave out the `streams` part as it seems no grouping / sharing of config options is needed for the synthetics use case.

My impression is that integrations packages do require the stream part, because this is what will generated when configuring an integration. If so, do we consciously make heartbeat unavailable for integrations authors here?

> The reason I prefer system/metrics system/logs is that it felt more intuitive to me to write and easier to remember

At the risk of derailing the discussion even more: We have no `system/logs` input. We have a `logfile` input for standalone users.

**Comment by ruflin** [comment]
We have the streamless concept for apm already, but this is only outside data streams. I wonder if this fits here too or not? As far as I know Synthetics team already built a package. Was it tested with this change?

On system/logs / logsfile, lets take this somewhere else as I agree it will derail further.

**Comment by andrewvc** [comment]
@ruflin @urso @justinkambic I've added `data_stream` support in my most recent commit 60bfdfecbf0bdd1f606f985b2b1821794564bf3f . I still need some sort of testing there, but it could probably use a quick review if @urso you have a minute to give your thoughts on doing it the way I'm doing. I looked at packetbeat's use of processors and did something similar.

**Comment by urso** [comment]
> I've added data_stream support in my most recent commit 60bfdfe . I still need some sort of testing there, but it could probably use a quick review if @urso you have a minute to give your thoughts on doing it the way I'm doing.

The commit doesn't seem to be included in this PR, but the idea looks good to me.

**Comment by andrewvc** [comment]
@urso apologies, pushed to the wrong remote, it's now updated.

File: heartbeat/monitors/factory.go [hunk 1]
```diff
@@ -52,9 +53,16 @@ type publishSettings struct {
 	// Output meta data settings
 	Pipeline string                   `config:"pipeline"` // ES Ingest pipeline name
 	Index    fmtstr.EventFormatString `config:"index"`    // ES output index pattern
+	DataStream *datastream `config:"data_stream"`
 	DataSet  string                   `config:"dataset"`
```
Comment (urso) [inline heartbeat/monitors/factory.go:57]: looks like you are missing a `go fmt`

File: heartbeat/monitors/factory.go [hunk 2]
```diff
@@ -84,8 +92,22 @@ func newCommonPublishConfigs(info beat.Info, cfg *common.Config) (pipetool.Confi
 	}
 
 	var indexProcessor processors.Processor
+	if settings.DataStream != nil {
+		index := fmt.Sprintf("%s-%s-%s",
+			settings.DataStream.Type,
+			settings.DataStream.Dataset,
+			settings.DataStream.Namespace)
```
Comment (urso) [inline heartbeat/monitors/factory.go:99]: if `Type` is not set you might want to set it to `synthetics`. Empty `Dataset` => `generic`, or empty `Namespace` would be `default`.

**Comment by urso** [comment]
Do we require modifications to the `heartbeat.spec`? Have you tested the modification with the synthetics package?

**Comment by andrewvc** [comment]
@urso we have not tested it yet, @justinkambic and @dominiqueclarke are working on the package and should be testing it against this branch shortly. I believe the [spec](https://github.com/elastic/beats/blob/master/x-pack/elastic-agent/spec/heartbeat.yml) should work fine here. Are there specific concerns you have?

File: heartbeat/monitors/factory.go (hunk 2 above)
Comment (andrewvc) [reply heartbeat/monitors/factory.go:99]: Hmmmm, I actually wonder if we should just hardcode the `type` is there actually use-case for overriding it given that heartbeat only ever produces the one type?

**Review by blakerouse — APPROVED** [review]
lgtm!

File: heartbeat/monitors/factory.go [hunk 3]
```diff
@@ -83,25 +92,24 @@ func newCommonPublishConfigs(info beat.Info, cfg *common.Config) (pipetool.Confi
...
 	userProcessors, err := processors.New(settings.Processors)
 	if err != nil {
 		return nil, err
 	}
 
+	// TODO: Remove this logic in the 8.0/master branch, preserve only in 7.x
 	dataset := settings.DataSet
 	if dataset == "" {
-		dataset = "uptime"
+		if settings.DataStream.Dataset != "" {
```
Comment (urso) [inline heartbeat/monitors/factory.go:108]: DataStream is a pointer. Do we need a `nil` check here?

File: heartbeat/monitors/factory.go [hunk 4]
```diff
@@ -140,3 +148,40 @@ func newCommonPublishConfigs(info beat.Info, cfg *common.Config) (pipetool.Confi
...
+			namespace = "default"
+		}
+		typ := settings.DataStream.Type
+		if typ == "" {
+			typ = "generic"
+		}
+
+		index := fmt.Sprintf("%s-%s-%s",
+			typ,
+			settings.DataStream.Dataset,
```
Comment (urso) [inline heartbeat/monitors/factory.go:166]: should top-level `Dataset` take precedence?

how about:

```
func defaultStr(str, def string) string {
  if str != "" {
    return str
  }
  return def
}

...

index := fmt.Sprintf("%s-%s-%s",
  defaultStr(settings.DataStream.Type, "synthetics"),
  defaultStr(settings.Dataset, defaultStr(settings.DataStream.Dataset, "generic")),
  defaultStr(settings.DataStream.Namespace, "default")
)
```

File: heartbeat/monitors/factory.go [hunk 5]
```diff
@@ -140,3 +148,40 @@ func newCommonPublishConfigs(info beat.Info, cfg *common.Config) (pipetool.Confi
...
+func setupIndexProcessor(info beat.Info, settings publishSettings) (processors.Processor, error) {
+	var indexProcessor processors.Processor
+	if settings.DataStream != nil {
+		namespace := settings.DataStream.Namespace
+		if namespace == "" {
+			namespace = "default"
+		}
+		typ := settings.DataStream.Type
+		if typ == "" {
+			typ = "generic"
```
Comment (urso) [inline heartbeat/monitors/factory.go:161]: Shouldn't this be `synthetics`, such that the index name becomes `synthetics-...` by default?

Comment (andrewvc) [reply heartbeat/monitors/factory.go:161]: Yes, I think I misunderstood a previous discussion of ours, glad to change it

File: heartbeat/monitors/factory.go (hunk 3 above)
Comment (andrewvc) [reply heartbeat/monitors/factory.go:108]: Good catch!

**Comment by andrewvc** [comment]
@urso I believe I've resolved your concerns in edf4173 , mind taking a look?

File: heartbeat/monitors/factory_test.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+func TestSetupIndexProcessor(t *testing.T) {
+	binfo := beat.Info{
+		Beat:        "heartbeat",
+		IndexPrefix: "heartbeat",
+		Version:     "8.0.0",
+	}
+	tests := []struct {
+		name          string
```
Comment (urso) [inline heartbeat/monitors/factory_test.go:38]: nit, remove name and:

```
tests := map[string]struct { ... }
```

**Comment by andrewvc** [comment]
@urso 92ea9ac should resolve your last concern re: test structure using a map

**Review by urso — APPROVED** [review]

**Pull request merged**